src/util/CacheableGenerator.h

```diff
@@ -0,0 +1,275 @@
+
+#ifndef CACHEABLEGENERATOR_H
+#define CACHEABLEGENERATOR_H
+
+#include <absl/cleanup/cleanup.h>
+
+#include <atomic>
+#include <chrono>
+#include <optional>
+#include <shared_mutex>
+#include <thread>
+#include <vector>
+
+#include "util/Exception.h"
+#include "util/Generator.h"
+#include "util/Synchronized.h"
+#include "util/UniqueCleanup.h"
+
+namespace ad_utility {
+
+class IteratorExpired : public std::exception {};
+
+template <typename T>
+class CacheableGenerator {
+  using GenIterator = typename cppcoro::generator<T>::iterator;
+
+  enum class MasterIteratorState { NOT_STARTED, MASTER_STARTED, MASTER_DONE };
+
+  class ComputationStorage {
+    friend CacheableGenerator;
+    mutable std::shared_mutex mutex_;
+    std::condition_variable_any conditionVariable_;
+    cppcoro::generator<T> generator_;
+    std::optional<GenIterator> generatorIterator_{};
+    std::vector<std::shared_ptr<T>> cachedValues_{};
+    MasterIteratorState masterState_ = MasterIteratorState::NOT_STARTED;
+    std::atomic<std::thread::id> currentOwningThread{};
+    // Returns true if cache needs to shrink, accepts a parameter that tells the
+    // callback if we actually can shrink, if the size changed because of a
+    // newly generated entry and the entry about to be removed or newly added.
+    std::function<bool(bool, bool, std::shared_ptr<const T>)> onSizeChanged_{};
+    std::function<void(std::chrono::milliseconds)> onNextChunkComputed_{};
+
+   public:
+    explicit ComputationStorage(cppcoro::generator<T> generator)
+        : generator_{std::move(generator)} {}
+    ComputationStorage(ComputationStorage&& other) = delete;
+    ComputationStorage(const ComputationStorage& other) = delete;
+    ComputationStorage& operator=(ComputationStorage&& other) = delete;
+    ComputationStorage& operator=(const ComputationStorage& other) = delete;
+
+   private:
+    void advanceTo(size_t index, bool isMaster) {
+      std::unique_lock lock{mutex_};
+      currentOwningThread = std::this_thread::get_id();
+      absl::Cleanup cleanup{
+          [this]() { currentOwningThread = std::thread::id{}; }};
+      AD_CONTRACT_CHECK(index <= cachedValues_.size());
+      // Make sure master iterator does exist and we're not blocking
+      // indefinitely
+      if (isMaster) {
+        AD_CORRECTNESS_CHECK(masterState_ != MasterIteratorState::MASTER_DONE);
+        masterState_ = MasterIteratorState::MASTER_STARTED;
+      } else {
+        AD_CORRECTNESS_CHECK(masterState_ != MasterIteratorState::NOT_STARTED);
+      }
+      if (index < cachedValues_.size()) {
+        if (!cachedValues_.at(index)) {
+          throw IteratorExpired{};
+        }
+        return;
+      }
+      if (generatorIterator_.has_value() &&
+          generatorIterator_.value() == generator_.end()) {
+        return;
+      }
+      if (masterState_ == MasterIteratorState::MASTER_STARTED && !isMaster) {
+        conditionVariable_.wait(lock, [this, index]() {
```

Comment:
I agree on the time limit, but as I told you in person the master/slave system does have some benefits so having only slaves pass a time limit would be an option, but this might be a task for a follow-up PR.

Reply:
I still think that the non-master approach + everybody having a timeout is better. We can probably dispute this further, the current state (forever blocking until the slow master times out) is a blocker for this PR.